src/scripts/analyze_missingness.py

```diff
@@ -0,0 +1,148 @@
+        then n(X) is used as a single value
+      - if some number of records 1, ..., j <= n have the same value
+        in X and all other records are blank in X, then the value
+        of the matching columns 1, ..., j is used as a singleton
+      - if some number of non-empty in X records 1, ..., j <= n have
+        different values in X, then all values 1(X), ..., j(X) are
+        concatenated into a list of values delimited by commas
+    """
+    non_nans = set([x for x in x.astype(str).to_list() if x != ""])
+    return ",".join(non_nans)
+
+
+def filter_for_valid_values(df: pd.DataFrame, values_by_column: dict[str, set[str]]):
+    """
+    Given a dataframe and a dictionary mapping columns to allowed values,
+    filter the df such that the provided columns contain only values
+    that are pre-defined as valid.
+    """
+    for col in values_by_column:
+        df[col] = df[col].map(lambda x: x if x in values_by_column[col] else "")
+    return df
+
+
+def count_patients(pre_linkage: pd.DataFrame, post_linkage: pd.DataFrame) -> dict:
+    """
+    Given a data frames with records before and after linkage return a dictionary
+    containing the number of distinct patients in each as well as the percent reduction.
+    """
+    patient_results = {}
+    patient_results["pre-linkage"] = len(pre_linkage)
+    patient_results["post-linkage"] = len(post_linkage)
+    patient_results["percent-reduction"] = round(
+        (
+            (patient_results["pre-linkage"] - patient_results["post-linkage"])
+            / patient_results["pre-linkage"]
+        )
+        * 100,
+        0,
+    )
+    return patient_results
+
+
+def count_records_missing_data(
+    pre_linkage: pd.DataFrame, post_linkage: pd.DataFrame, columns_and_values: dict
+) -> dict:
+    """
+    Given datasets pre and post linkage as well as a dictionary whose keys specify
+    columns of interests and values are sets of acceptable values for the corresponding
+    columns return a dictionary reporting the number of records missing data in each of
+    the columns of interest.
+    """
+    results = {
+        "field": [],
+        "pre-linkage-count": [],
+        "pre-linkage-percent": [],
+        "post-linkage-count": [],
+        "post-linkage-percent": [],
+    }
+    for idx, column in enumerate(columns_and_values):
+        results["field"].append(column)
+        results["pre-linkage-count"].append(
+            len(pre_linkage.loc[~pre_linkage[column].isin(columns_and_values[column])])
+        )
+        results["pre-linkage-percent"].append(
+            round((results["pre-linkage-count"][idx] / len(pre_linkage)) * 100, 1)
+        )
+        results["post-linkage-count"].append(
+            len(post_linkage.loc[post_linkage[column].apply(len) == 0])
+        )
+        results["post-linkage-percent"].append(
+            round((results["post-linkage-count"][idx] / len(post_linkage)) * 100, 1)
+        )
+    return results
```

Comment:
So this is what I think of as Dan's quick and dirty solution to calculating four numbers quickly. Because records have been linked in the various data frames, he can use the length of the data frame (just the number of rows) as an accurate proxy for the number of people in the data set. So it's in essence just a bunch of number grabs with some math formatting and rounding. The `.loc`'s are for getting real data versions of the data frames instead of just copied views (which misbehave like always).